README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# aquodom

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/aquodom)](https://CRAN.R-project.org/package=aquodom)
[![R-CMD-check](https://github.com/RedTent/aquodom/workflows/R-CMD-check/badge.svg)](https://github.com/RedTent/aquodom/actions)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

Met *aquodom* is het op eenvoudige wijze mogelijk om de
aquo-domeintabellen te downloaden en te gebruiken in R.

## Installatie

*aquodom* is te installeren vanaf CRAN met:

``` r
install.packages("aquodom") 
```

De ontwikkelversie is te installeren van
[GitHub](https://github.com/RedTent/aquodom) met:

``` r
# install.packages("devtools")
devtools::install_github("RedTent/aquodom")
```

## Aquo-domeintabellen

De Aquo-standaard vormt de Nederlandse standaard voor de uitwisseling
van gegevens in het waterbeheer. Met *aquodom* (kort voor
aquo-domeintabellen) is het makkelijk om via de API domeintabellen van
de Aquo-standaard in R te downloaden en te gebruiken.

De belangrijkste functies van *aquodom* zijn `dom()` en `dom_save()`.
Met deze functies kan iedere domeintabel van www.aquo.nl worden
gedownload. De functie `dom()` geeft de domeintabel als een dataframe.
`dom_save()` doet hetzelfde maar slaat daarnaast ook de domeintabel op
als xlsx- of csv-bestand. De functie `dom_overzicht()` geeft een
compleet overzicht van alle beschikbare domeintabellen.

Alle functies hebben een optioneel argument `peildatum`. Dit argument
kan worden gebruikt om alleen domeinwaarden of domeintabellen te tonen
die geldig zijn op de peildatum. Met `peildatum = NULL` worden alle
resultaten inclusief verouderde waarden getoond.

``` r
library(aquodom)

dom("MonsterType")
#> # A tibble: 7 x 6
#>      id omschrijving   begin_geldigheid eind_geldigheid guid      gerelateerd   
#>   <dbl> <chr>          <date>           <date>          <chr>     <chr>         
#> 1    10 analysemonster 2017-12-13       2100-01-01      Id-6a3e6~ Id-99092d94-d~
#> 2     8 materiaalmons~ 2015-11-18       2100-01-01      Id-f811d~ Id-2d146a3e-3~
#> 3    11 samengevoegd ~ 2017-12-13       2100-01-01      Id-81ce3~ Id-8df42796-7~
#> 4     4 toetsmonster   2015-11-18       2100-01-01      Id-0034d~ Id-ad4f1180-6~
#> 5     9 uitloogmonster 2015-11-18       2100-01-01      Id-6053f~ Id-48826f74-c~
#> 6     1 veldmonster    2015-11-18       2100-01-01      Id-74dd8~ Id-3e9918e3-4~
#> 7     7 zeefmonster    2015-11-18       2100-01-01      Id-8d483~ Id-63ac95ff-1~

#De namen van domeintabellen zijn niet hoofdlettergevoelig
all.equal(dom("MonsterType"), dom("monstertype"))
#> [1] TRUE

# Alle domeinwaarden inclusief verouderde waarden
dom("MonsterType", peildatum = NULL)
#> # A tibble: 8 x 6
#>      id omschrijving   begin_geldigheid eind_geldigheid guid      gerelateerd   
#>   <dbl> <chr>          <date>           <date>          <chr>     <chr>         
#> 1    10 analysemonster 2017-12-13       2100-01-01      Id-6a3e6~ Id-99092d94-d~
#> 2     8 materiaalmons~ 2015-11-18       2100-01-01      Id-f811d~ Id-2d146a3e-3~
#> 3    10 mengmonster    2015-11-18       2017-12-12      Id-a0a78~ <NA>          
#> 4    11 samengevoegd ~ 2017-12-13       2100-01-01      Id-81ce3~ Id-8df42796-7~
#> 5     4 toetsmonster   2015-11-18       2100-01-01      Id-0034d~ Id-ad4f1180-6~
#> 6     9 uitloogmonster 2015-11-18       2100-01-01      Id-6053f~ Id-48826f74-c~
#> 7     1 veldmonster    2015-11-18       2100-01-01      Id-74dd8~ Id-3e9918e3-4~
#> 8     7 zeefmonster    2015-11-18       2100-01-01      Id-8d483~ Id-63ac95ff-1~

head(dom_overzicht(), 3)
#> # A tibble: 3 x 7
#>   domeintabel domeintabelsoort wijzigingsdatum begin_geldigheid eind_geldigheid
#>   <chr>       <chr>            <date>          <date>           <date>         
#> 1 Afsluitmid~ Domeintabel      2020-11-11      2016-03-12       2100-01-01     
#> 2 Bekleding_~ Domeintabel      2020-06-30      2016-03-12       2100-01-01     
#> 3 BekledingT~ Domeintabel      2020-06-30      2016-03-12       2100-01-01     
#> # ... with 2 more variables: kolommen <list>, guid <chr>

nrow(dom_overzicht())
#> [1] 126
# inclusief ongeldige domeintabellen
nrow(dom_overzicht(peildatum = NULL))
#> [1] 261
```

``` r
dom_save("monstertype")
```

## Caching en opslaan

Het downloaden van grotere domeintabellen kan behoorlijk wat tijd in
beslag nemen. Daarom maakt *aquodom* gebruik van caching. Als een
domeintabel eenmaal is gedownload wordt in dezelfde R-sessie gebruik
gemaakt van de cache. Voor het gebruik van dezelfde domeintabel in
verschillende R-sessies kan de domeintabel het beste opgeslagen worden,
bijv. met `dom_save()`.

``` r
# De eerste keer duurt vrij lang
system.time(dom("Hoedanigheid"))
#> ..
#>    user  system elapsed 
#>    0.27    0.14    2.06

# De tweede keer gaat veel sneller
system.time(dom("Hoedanigheid"))
#>    user  system elapsed 
#>    0.02    0.00    0.02
```


NEWS.md
# aquodom 0.1.1

## Bugfixes

* Nieuwe readr-versie gebruikt andere name-repair wat leidde tot foutmeldingen. Deze fout is verholpen zodat aquodom ook met de nieuwere versies van readr werkt.
* Foutmelding bij ongeldige naam domeintabel verbeterd. (#1)
* Fout opgelost waarbij oude domeintabellen niet opgevraagd konden worden (#2)
* Fout opgelost met datumconversie als gevolg van verschillende datumformats bij Aquo.

# aquodom 0.1.0

* `dom()` - Vraag een domeintabel op.
* `dom_save()` - Opslaan van een domeintabel
* `dom_overzicht()` - Overzicht van alle domeintabellen.



tests/testthat.R
library(testthat)
library(aquodom)

test_check("aquodom")


tests/testthat/test-dom.R
test_that("dom werkt", {
  skip_on_ci()
  skip_on_cran()


  expect_error(dom("test"), "geen geldige")
  monstertype <- dom("MonsterType")
  expect_s3_class(monstertype$begin_geldigheid, "Date")
  expect_s3_class(monstertype$eind_geldigheid, "Date")
  expect_equal(names(monstertype), c("id", "omschrijving", "begin_geldigheid", "eind_geldigheid", "guid","gerelateerd"))

  monstertype <- dom("MonsterType", peildatum = "2021-04-03")
  expect_equal(nrow(monstertype), 7)
  monstertype <- dom("MonsterType", peildatum = as.Date("2021-04-03"))
  expect_equal(nrow(monstertype), 7)

  monstertype <- dom("monstertype", peildatum = as.Date("2021-04-03"))
  expect_equal(nrow(monstertype), 7)

  expect_lt(system.time(dom("MonsterType"))[[3]], 0.5)

})


tests/testthat/test-dom_helpers.R
test_that("is_domeintabel",{
  skip_on_ci()
  skip_on_cran()

  expect_equal(is_domeintabel(c("Hoedanigheid", "Domeintabel")), c(TRUE, FALSE))
  expect_true(is_domeintabel("Academische_titel"))
  expect_true(is_domeintabel("monstertype"))
})

test_that("dom_guid",{
  skip_on_ci()
  skip_on_cran()

  expect_equal(dom_guid(c("Hoedanigheid", "Domeintabel")), c("Id-7169dd0a-813b-4cf1-86ab-9bbc52b113a4", NA))
  expect_equal(dom_guid("hoedanigheid"), c("Id-7169dd0a-813b-4cf1-86ab-9bbc52b113a4"))
})

test_that("dom_kolommen",{
  skip_on_ci()
  skip_on_cran()

  expect_equal(dom_kolommen("MonsterType"), c("Begin geldigheid", "Eind geldigheid", "Gerelateerd", "Id", "Omschrijving"))
  expect_equal(dom_kolommen("monstertype"), c("Begin geldigheid", "Eind geldigheid", "Gerelateerd", "Id", "Omschrijving"))
})


R/dom_save.R
#' Opslaan domeintabel
#'
#' Met deze functie is het mogelijk om een domeintabel op te slaan als .xlsx-bestand of als .csv-bestand.
#'
#' @inheritParams dom
#'
#' @param bestandsnaam Naam van het bestand om op te slaan. Eventueel als volledig pad. Default is de datum
#' gevolgd door de naam van de domeintabel. De bestandsnaam mag zonder extensie worden opgegeven.
#' @param map Naam van de map. De map moet bestaan. Optioneel.
#' @param bestandstype "xlsx" of "csv"  Opslaan als xlsx of als csv-bestand. Default is xlsx.
#'
#' @return Slaat de domeintabel op de schijf. De domeintabel zelf wordt onzichtbaar geretourneerd.
#'
#' @details
#' Voor het opslaan van een bestand als .xlsx wordt [openxlsx::write.xlsx()] gebruikt. Voor het opslaan
#' als .csv wordt [readr::write_csv2()] gebruikt.
#'
#' Opgeslagen excelbestanden kunnen ingelezen worden met [openxlsx::read.xlsx()] of [readxl::read_excel()].
#' csv-bestanden kunnen worden ingelezen met [readr::read_csv2()]
#'
#' @family domeintabellen
#'
#' @export
#'
#' @examples
#'
#' \dontrun{
#'  dom_save("MonsterType")
#'  dom_save("MonsterType", bestandsnaam = "test.csv", map = "data", bestandstype = "csv")
#'
#' }
dom_save <- function(naam,
                     bestandsnaam = paste(Sys.Date(), naam),
                     map = NULL,
                     bestandstype = c("xlsx", "csv"),
                     peildatum = Sys.Date()) {

  bestandstype <- bestandstype[[1]]
  rlang::arg_match(bestandstype, c("xlsx", "csv"))

  dom_tabel <- dom(naam, peildatum)

  if (bestandstype == "xlsx") {
    bestandsnaam <- bestandsnaam %>% stringr::str_remove(".xlsx$") %>% stringr::str_c(".xlsx")
    if (!is.null(map)) bestandsnaam <- file.path(map, bestandsnaam)

    openxlsx::write.xlsx(dom_tabel, file = bestandsnaam)
  }

  if (bestandstype == "csv") {
    bestandsnaam <- bestandsnaam %>% stringr::str_remove(".csv$") %>% stringr::str_c(".csv")
    if (!is.null(map)) bestandsnaam <- file.path(map, bestandsnaam)

    readr::write_csv2(dom_tabel, path = bestandsnaam)
  }

  invisible(dom_tabel)

}

# dom_save("MonsterType", "DEV/test.xlsx", bestandstype = "xlsx")
# x <- dom_save("MonsterType", "DEV/test.xlsx")
# file.remove("DEV/test.xlsx")
#
# dom_save("MonsterType", "DEV/test.csv", "csv", "20000101")
# file.remove("DEV/test.csv")
# dom_save("MonsterType")
# file.remove(paste(Sys.Date(),"MonsterType.xlsx"))
# dom_save("MonsterType", bestandstype = "csv")
# file.remove(paste(Sys.Date(),"MonsterType.csv"))
# dom_save("MonsterType", map = "DEV")
# file.remove(file.path("DEV", paste(Sys.Date(),"MonsterType.xlsx")))
#
# naam <- "MonsterType"


R/dom_helpers.R
#' Check domeintabelnamen
#'
#' Checkt of een namen geldige domeintabelnamen zijn.
#'
#' @param namen Character vector met namen van domeintabellen.
#'
#' @return Logical vector
#'
#' @export
#'
#' @examples
#' \dontrun{
#'
#' is_domeintabel("MonsterType")
#' is_domeintabel("Domeintabel")
#'
#' }
is_domeintabel <- function(namen){
  overzicht <- dom_overzicht(peildatum = NULL)
  stringr::str_to_lower(namen) %in% stringr::str_to_lower(overzicht$domeintabel)
}

#' Guid van domeintabel
#'
#' Zoek de guid van domeintabellen op.
#'
#' @param namen Character vector met namen van domeintabellen.
#'
#' @return Vector met guid's
#'
#' @export
#'
#' @examples
#' \dontrun{
#'
#' dom_guid("MonsterType")
#'
#' }
dom_guid <- function(namen){

  namen <- dom_convert_case(namen)

  overzicht <- dom_overzicht(peildatum = NULL)
  tibble::tibble(namen = namen) %>%
    dplyr::left_join(overzicht, by = c("namen" = "domeintabel")) %>%
    dplyr::pull(guid) %>%
    unname()
}


#' Kolommen van een domeintabel
#'
#' Deze functie zoekt op welke kolommen een domeintabel heeft.
#'
#' @param naam Naam van een domeintabel
#'
#' @return Een vector met kolomnamen
#'
#' @export
#'
#' @examples
#' \dontrun{
#'
#' dom_kolommen("MonsterType")
#'
#' }
dom_kolommen <- function(naam){
  if (length(naam) != 1) stop("`naam` dient een vector met lengte 1 te zijn")

  if (!is_domeintabel(naam)) stop(paste(naam, "is geen geldige domeintabelnaam"))

  overzicht <- dom_overzicht(peildatum = NULL)

  naam <- dom_convert_case(naam)

  overzicht %>%
    dplyr::filter(domeintabel == naam) %>%
    dplyr::pull(kolommen) %>%
    .[[1]]
}

#' Maak een URL naar de csv domeintabel
#'
#' Deze functie genereert de juiste URL van een domeintabel voor de functie `dom()`
#'
#' @param naam Naam van de domeintabel
#' @param limit Aantal waarden per keer
#' @param offset Startpunt
#'
#' @return Een string met de URL
#'
#' @noRd
create_dom_url <- function(naam, limit = 500, offset = 0){

  kolomstring <- dom_kolommen(naam) %>%
    stringr::str_replace(" ", "+") %>%
    paste(collapse = "%0D%0A?")

  glue::glue(
    "https://www.aquo.nl/index.php?title=Speciaal:Vragen&q=+[[Breder::{dom_guid(naam)}]]",
    "%0D%0A&po=?{kolomstring}%0D%0A",
    "&p[format]=csv&p[sep]=;&p[limit]={limit}&p[offset]={offset}"
  )
}

#' Conversie van hoofdletters van domeintabel namen
#'
#' @param namen
#'
#' @return `namen` met de juiste hoofdletters.
#'
#' @noRd
#'
dom_convert_case <- function(namen) {

  opzoektabel <-
    dom_overzicht(peildatum = NULL) %>%
    dplyr::mutate(lower = stringr::str_to_lower(domeintabel)) %>%
    dplyr::select(lower, domeintabel) %>%
    tibble::deframe()

  unname(opzoektabel[as.character(stringr::str_to_lower(namen))])
}


R/dom_overzicht.R
#' Overzicht van alle domeintabellen
#'
#' Deze functie geeft een overzicht van alle beschikbare domeintabellen,
#' inclusief historische tabellen, op www.aquo.nl.
#'
#' @param peildatum Date of een character die omgezet kan worden in een Date met
#'   `lubridate::as_date()`. De peildatum filtert de output om alleen geldige
#'   domeintabellen op de peildatum weer te geven. Gebruik `NULL` om alle
#'   domeintabellen ongeacht de geldigheid weer te geven.
#'
#' @section Caching: Deze functie maakt gebruik van caching voor het
#'   optimaliseren van snelheid en om de aquo-server niet onnodig te belasten.
#'   Hiervoor wordt de map `tempdir()` gebruikt als cache. Deze map wordt na
#'   elke R-sessie verwijderd.
#'
#' @return Een tibble met een overzicht van alle domeintabellen. Het overzicht
#'   bevat de volgende kolommen:
#'
#'   - `domeintabel` - Naam van de domeintabel.
#'   - `domeintabelsoort` - Het type domeintabel.
#'   - `wijzigingsdatum` - Datum van de laatste wijziging van de tabel.
#'   - `begin_geldigheid` - Datum van het begin van de geldigheid van de domeintabel.
#'   - `eind_geldigheid` - Datum van het eind van de geldigheid van de domeintabel.
#'   - `kolommen` - Een vector met de kolomnamen van de domeintabel.
#'   - `guid` - De guid van de domeintabel.
#'
#' @family domeintabellen
#'
#' @export
#'
#' @examples
#' \dontrun{
#'
#' dom_overzicht()
#' dom_overzicht(peildatum = Sys.Date())
#' dom_overzicht(peildatum = "2021-04-05")
#'
#' }
dom_overzicht <- function(peildatum = Sys.Date()){

  dom_overzicht_m <- memoise::memoise(dom_overzicht_basis,
                                      cache = cachem::cache_disk(dir = tempdir()))

  overzicht <- suppressWarnings(dom_overzicht_m())

  if (!is.null(peildatum)) {
    if (class(peildatum) != "Date") {peildatum <- lubridate::as_date(peildatum)}
    overzicht <- overzicht %>% dplyr::filter(begin_geldigheid <= peildatum, eind_geldigheid >= peildatum)
  }
  return(overzicht)
}

dom_overzicht_basis <- function() {

  url <- "https://www.aquo.nl/index.php?title=Speciaal:Vragen&x=[[Elementtype::Domeintabel%20%7C%7C%20Domeintabeltechnisch%20%7C%7C%20Domeintabelverzamellijst]]%20/?Elementtype/?Voorkeurslabel/?Metadata/?Wijzigingsdatum/?Begin%20geldigheid/?Eind%20geldigheid/&format=csv&sep=;&offset=0&limit=500"

  req <- httr::GET(url)

  if (req$status_code != 200 || length(req$content) == 0) {
    message("Geen domeintabellen gevonden")
    return(NULL)
  }

  datum_formats <- c("dBY HMS", "dBY", "dmY HMS", "Ymd HMS", "Ymd", "dmY")

  overzicht <- req %>%
    httr::content(as = "text") %>%
    readr::read_csv2(locale = readr::locale(decimal_mark = ",", grouping_mark = ".")) %>%
    dplyr::rename_with(.fn = ~"guid", .cols = dplyr::any_of(c("X1", "...1"))) %>% # ivm verschillen in readr versies
    dplyr::select(domeintabel = Voorkeurslabel,
                  domeintabelsoort = Elementtype,
                  wijzigingsdatum = Wijzigingsdatum,
                  begin_geldigheid = `Begin geldigheid`,
                  eind_geldigheid = `Eind geldigheid`,
                  kolommen = Metadata,
                  guid) %>%
    dplyr::mutate(dplyr::across(.cols = c(wijzigingsdatum, begin_geldigheid, eind_geldigheid),
                                .fns = ~lubridate::as_date(lubridate::parse_date_time(.x, orders = datum_formats)))
                                  ) %>%
    dplyr::mutate(kolommen = stringr::str_split(kolommen, ","))

  return(overzicht)
}






R/dom.R
#' Opvragen domeintabel
#'
#' Deze functie haalt een domeintabel op van www.aquo.nl.
#'
#' @param naam Naam van een domeintabel - De namen zijn niet hoofdlettergevoelig. Zie
#'   `dom_overzicht()` voor geldige domeintabelnamen.
#' @param peildatum Date of een character die omgezet kan worden in een Date met
#'   `lubridate::as_date()`. De peildatum filtert de output om alleen geldige
#'   domeinwaarden op de peildatum weer te geven. Gebruik `NULL` om alle
#'   domeinwaarden ongeacht de geldigheid weer te geven.
#'
#' @section Caching: Deze functie maakt gebruik van caching voor het
#'   optimaliseren van snelheid en om de aquo-server niet onnodig te belasten.
#'   Hiervoor wordt de map `tempdir()` gebruikt als cache. Deze map wordt na
#'   elke R-sessie verwijderd.
#'
#' @return Een tibble met een met domeinwaarden. De kolommen zijn afhankelijk
#'   van de betreffende domeintabel.
#'
#' @family domeintabellen
#'
#' @export
#'
#' @examples
#' \dontrun{
#'
#' dom("MonsterType")
#' dom("MonsterType", peildatum = Sys.Date())
#' dom("MonsterType", peildatum = "2021-04-05")
#'
#' }
dom <- function(naam, peildatum = Sys.Date()) {

  if (length(naam) != 1) stop("`naam` dient een vector met lengte 1 te zijn")
  if (!is_domeintabel(naam)) stop(paste(naam, "is geen geldige domeintabelnaam"))

  naam <- dom_convert_case(naam)

  dom_m <- memoise::memoise(dom_basis, cache = cachem::cache_disk(dir = tempdir()))

  domeintabel <- suppressWarnings(dom_m(naam))

  if (!is.null(peildatum)) {
    if (!"begin_geldigheid" %in% names(domeintabel) | !"eind_geldigheid" %in% names(domeintabel)) {
      stop("Voor deze tabel is geen begin_geldigheid of eind_geldigheid beschikbaar")
    }
    if (class(peildatum) != "Date") {peildatum <- lubridate::as_date(peildatum)}
    domeintabel <- domeintabel %>% dplyr::filter(begin_geldigheid <= peildatum, eind_geldigheid >= peildatum)

  }

  return(domeintabel)
}

dom_basis <- function(naam){

  limit <- 500
  offset <- 0
  res <- tibble::tibble()
  while(TRUE){ # break out with break

    dom_deel_url <- create_dom_url(naam, limit, offset)

    deel_res <-
      dom_deel_url %>%
      httr::GET() %>%
      httr::content(as = "text") %>%
      readr::read_csv2(locale = readr::locale(decimal_mark = ",", grouping_mark = "."))

    res <- dplyr::bind_rows(res, deel_res)

    if (nrow(deel_res) < limit) break()

    offset <- offset + limit
    cat(".")
  }
  cat("\n")

  datum_formats <- c("dBY HMS", "dBY", "dmY HMS", "Ymd HMS", "Ymd", "dmY")

  res <-
    res %>%
    dplyr::rename_with(.fn = ~"guid", .cols = dplyr::any_of(c("...1", "X1"))) %>%
    dplyr::rename_with(.fn = function(x) stringr::str_replace(x, pattern = " ", "_")) %>%
    dplyr::rename_with(.fn = stringr::str_to_lower) %>%
    dplyr::mutate(dplyr::across(.cols = dplyr::contains(c("geldigheid", "datum")),
                                .fns = ~lubridate::as_date(lubridate::parse_date_time(.x, orders = datum_formats)))
                  ) %>%
    dplyr::relocate(dplyr::any_of(c("id", "codes", "cijfercode", "omschrijving", "begin_geldigheid", "eind_geldigheid")))

  return(res)
}


